pkg/apis/flowcontrol/types.go

```diff
@@ -0,0 +1,228 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package flowcontrol
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// These are valid flow-dingtinguisher methods.
+const (
+	// FlowDistinguisherMethodByUser works by computing flow distinguisher with the user name in the request context so
+	// that the requests from the same user will enqueued into the same queue for processing.
+	FlowDistinguisherMethodByUser FlowDistinguisherMethodType = "ByUser"
```

Comment:
(1) Sadly, since "Type" is in the type name it should also be in the value name.

(2) We need `FlowDistinguisherMethodTypeNone = ""`.

Reply:
> (1) Sadly, since "Type" is in the type name it should also be in the value name.

added "Type" in their names.

> (2) We need FlowDistinguisherMethodTypeNone = "".

mmm not fully convinced of the point. it will the same if we set the `queues` to 1, isn't it? additionally, what'll happen if both a none-method flow-schema and by-xxx-method flow-schema references a same PL? i think it'll lead to more complexity wrt https://github.com/kubernetes/kubernetes/pull/77048#discussion_r285422132

Reply:
(2a) We already have the possibility of different FlowDistinguisherMethods being used with the same priority level --- and that is not a problem.  All FlowDistinguisherMethods prescribe a uniform thing --- how to compute the flow distinguisher string.

(2b) I think that for ease of use we should allow the zero value of `FlowDistinguisherMethod`.  That means we should allow the method type value to be `""`.  Since that is now an enumerated type, we should define `""` as a member of that enumerated type.

Reply:
Oh, wait, I was mistaken about the rendering of a struct-typed field that has its zero value.  I thought it was completely omitted from the YAML rending, but that is wrong; it is rendered as `{}`.  So I think it is right to type this field as a pointer to a struct and use the `nil` value to specify the degenerate flow distinguisher method.  The latest revision does that, so I think we can mark this conversation as resolved.